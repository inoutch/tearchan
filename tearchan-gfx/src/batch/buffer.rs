use crate::buffer::BufferTrait;
use std::cmp::Ordering;
use std::collections::{BTreeMap, VecDeque};
use std::hash::{Hash, Hasher};
use std::marker::PhantomData;
use tearchan_util::btree::DuplicatableBTreeMap;

#[derive(Debug)]
pub enum BatchBufferAllocatorEvent {
    Write(BatchBufferPointer),
    Clear(BatchBufferPointer),
    ReallocateAll {
        pairs: Vec<BatchBufferAllocatorReallocPair>,
    },
}

#[derive(Debug)]
pub struct BatchBufferAllocatorReallocPair {
    pub from: BatchBufferPointer,
    pub to: BatchBufferPointer,
}

#[derive(Debug)]
enum Event {
    Clear {
        pointer: BatchBufferPointer,
    },
    Write {
        first: usize,
    },
    Reallocate {
        pairs: Vec<BatchBufferAllocatorReallocPair>,
    },
}

#[derive(Default)]
pub struct BatchBufferAllocator {
    pointers: BTreeMap<usize, BatchBufferPointer>, // grouped by first
    pending_pointers: DuplicatableBTreeMap<usize, usize>, // grouped by size, value is last of pending pointer
    pending_pointers_grouped_by_last: BTreeMap<usize, BatchBufferPointer>,
    events: VecDeque<Event>,
    len: usize,
}

impl BatchBufferAllocator {
    pub fn allocate(&mut self, len: usize) -> BatchBufferPointer {
        if let Some(pointer) = self.allocate_from_pending_pointers(len) {
            self.events.push_back(Event::Write {
                first: pointer.first,
            });
            return pointer;
        }
        let first = self
            .pointers
            .last_key_value()
            .map(|(key, pointer)| *key + pointer.len)
            .unwrap_or(0);
        let pointer = BatchBufferPointer::new(first, len);
        self.pointers.insert(first, pointer);
        self.len += len;
        self.events.push_back(Event::Write { first });
        pointer
    }

    pub fn reallocate(&mut self, pointer: BatchBufferPointer, size: usize) -> BatchBufferPointer {
        self.free(pointer);
        self.allocate(size)
    }

    pub fn free(&mut self, pointer: BatchBufferPointer) {
        let first = pointer.first;
        let last = pointer.first + pointer.len;
        self.pointers.remove(&pointer.first);

        let is_last = first == self.len - pointer.len;
        let mut seek = first;

        let mut merge_pointers = vec![pointer];
        for (last, pointer) in self.pending_pointers_grouped_by_last.range(0..=first).rev() {
            if &seek != last {
                break;
            }
            merge_pointers.push(*pointer);
            seek -= pointer.len;
        }

        for merge_pointer in merge_pointers {
            let last = merge_pointer.first + merge_pointer.len;
            self.pending_pointers.remove(&merge_pointer.len, &last);
            self.pending_pointers_grouped_by_last.remove(&last);
        }

        if is_last {
            self.len = seek;
            return;
        }

        self.push_pending_pointer(BatchBufferPointer::new(seek, last - seek));
        self.events.push_back(Event::Clear { pointer });
    }

    pub fn defragmentation(&mut self) {
        self.pending_pointers.clear();
        self.pending_pointers_grouped_by_last.clear();
        self.events.clear();

        let prev_pointers = std::mem::take(&mut self.pointers);
        let mut seek = 0;
        let mut pairs = Vec::new();
        let mut write_events = VecDeque::new();
        for (_, pointer) in prev_pointers.into_iter() {
            let len = pointer.len;
            let new_pointer = BatchBufferPointer::new(seek, len);
            self.pointers.insert(seek, new_pointer);

            write_events.push_back(Event::Write {
                first: new_pointer.first,
            });
            pairs.push(BatchBufferAllocatorReallocPair {
                from: pointer,
                to: new_pointer,
            });

            seek += len;
        }

        self.events.push_back(Event::Reallocate { pairs });
        self.events.append(&mut write_events);

        self.len = seek;
    }

    pub fn pop_event(&mut self) -> Option<BatchBufferAllocatorEvent> {
        match self.events.pop_front()? {
            Event::Write { first } => {
                let pointer = self.pointers.get(&first)?;
                Some(BatchBufferAllocatorEvent::Write(*pointer))
            }
            Event::Clear { pointer } => Some(BatchBufferAllocatorEvent::Clear(pointer)),
            Event::Reallocate { pairs } => Some(BatchBufferAllocatorEvent::ReallocateAll { pairs }),
        }
    }

    pub fn len(&self) -> usize {
        self.len
    }

    pub fn is_empty(&self) -> bool {
        self.len == 0
    }

    pub fn sort_by<F>(&mut self, compare: F)
    where
        F: FnMut(&BatchBufferPointer, &BatchBufferPointer) -> Ordering,
    {
        let mut pointers = self.pointers.iter().map(|(_, p)| *p).collect::<Vec<_>>();
        pointers.sort_by(compare);

        self.events.clear();
        self.pointers.clear();
        self.pending_pointers.clear();
        self.pending_pointers_grouped_by_last.clear();

        let mut pairs = Vec::new();
        let mut write_events = VecDeque::new();
        let mut seek = 0;
        for pointer in pointers {
            let new_pointer = BatchBufferPointer::new(seek, pointer.len);
            self.pointers.insert(new_pointer.first, new_pointer);

            write_events.push_back(Event::Write {
                first: new_pointer.first,
            });
            pairs.push(BatchBufferAllocatorReallocPair {
                from: pointer,
                to: new_pointer,
            });

            seek += new_pointer.len;
        }

        self.events.push_back(Event::Reallocate { pairs });
        self.events.append(&mut write_events);

        assert!(seek <= self.len, "{} <= {}", seek, self.len);
        self.len = seek;
    }

    fn allocate_from_pending_pointers(&mut self, len: usize) -> Option<BatchBufferPointer> {
        let last = self
            .pending_pointers
            .range_mut(len..)
            .next()?
            .1
            .pop_front()?;
        let pointer = self.pending_pointers_grouped_by_last.remove(&last).unwrap();
        {
            let (pointer, pending_pointer) = if len == pointer.len {
                (pointer, None)
            } else {
                (
                    BatchBufferPointer::new(pointer.first, len),
                    Some(BatchBufferPointer::new(
                        pointer.first + len,
                        pointer.len - len,
                    )),
                )
            };
            self.pointers.insert(pointer.first, pointer);
            if let Some(pending_pointer) = pending_pointer {
                self.push_pending_pointer(pending_pointer);
            }
            Some(pointer)
        }
    }

    fn push_pending_pointer(&mut self, pointer: BatchBufferPointer) {
        let last = pointer.first + pointer.len;
        self.pending_pointers_grouped_by_last.insert(last, pointer);
        self.pending_pointers.push_front(pointer.len, last);
    }
}

#[derive(Copy, Clone, Debug)]
pub struct BatchBufferPointer {
    first: usize,
    len: usize,
}

impl Eq for BatchBufferPointer {}

impl PartialEq for BatchBufferPointer {
    fn eq(&self, other: &Self) -> bool {
        self.first == other.first
    }
}

impl Ord for BatchBufferPointer {
    fn cmp(&self, other: &Self) -> Ordering {
        self.first.cmp(&other.first)
    }
}

impl PartialOrd for BatchBufferPointer {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        self.first.partial_cmp(&other.first)
    }
}

impl Hash for BatchBufferPointer {
    fn hash<H: Hasher>(&self, state: &mut H) {
        state.write_usize(self.first)
    }
}

impl BatchBufferPointer {
    fn new(first: usize, len: usize) -> Self {
        BatchBufferPointer { first, len }
    }

    pub fn first(&self) -> usize {
        self.first
    }
}

pub struct BatchBuffer<TBuffer, TDataType> {
    buffer: TBuffer,
    _phantom: PhantomData<TDataType>,
}

impl<'a, TBuffer, TDataType> BatchBuffer<TBuffer, TDataType>
where
    TBuffer: BufferTrait<'a, TDataType>,
{
    pub fn new(buffer: TBuffer) -> Self {
        Self {
            buffer,
            _phantom: PhantomData,
        }
    }

    pub fn write(
        &mut self,
        writer: TBuffer::Writer,
        pointer: BatchBufferPointer,
        data: &[TDataType],
    ) {
        assert!(pointer.len <= data.len());
        self.buffer.write(writer, data, pointer.first);
    }

    pub fn clear(&mut self, writer: TBuffer::Writer, pointer: BatchBufferPointer) {
        self.buffer.clear(writer, pointer.first, pointer.len);
    }

    pub fn resize(&mut self, resizer: TBuffer::Resizer, len: usize) {
        self.buffer.resize(resizer, len);
    }

    #[inline]
    pub fn len(&self) -> usize {
        self.buffer.len()
    }

    #[inline]
    pub fn is_empty(&self) -> bool {
        self.buffer.is_empty()
    }

    pub fn buffer(&self) -> &TBuffer {
        &self.buffer
    }
}

#[cfg(test)]
mod test {
    use crate::batch::buffer::{
        BatchBuffer, BatchBufferAllocator, BatchBufferAllocatorEvent, BatchBufferPointer,
    };
    use crate::buffer::BufferTrait;
    use std::collections::HashMap;

    struct ResultContext;

    #[derive(Default)]
    struct VecBuffer {
        data: Vec<u32>,
    }

    impl VecBuffer {
        pub fn new(len: usize) -> VecBuffer {
            VecBuffer { data: vec![0; len] }
        }
    }

    impl<'a> BufferTrait<'a, u32> for VecBuffer {
        type Resizer = &'a mut ResultContext;
        type Writer = &'a mut ResultContext;
        type Copier = &'a mut ResultContext;

        fn resize(&mut self, _resizer: Self::Resizer, len: usize) {
            self.data.resize(len, 0);
        }

        fn write(&mut self, _writer: &mut ResultContext, data: &[u32], offset: usize) {
            self.data
                .splice(offset..(offset + data.len()), data.iter().copied());
        }

        fn copy(&mut self, _copy: &mut ResultContext, from: usize, to: usize, len: usize) {
            let from = {
                self.data.as_slice()[from..(from + len)]
                    .iter()
                    .copied()
                    .collect::<Vec<_>>()
            };
            self.data.splice(to..(to + len), from);
        }

        fn len(&self) -> usize {
            self.data.len()
        }

        fn is_empty(&self) -> bool {
            self.data.is_empty()
        }

        fn clear(&mut self, _writer: &mut ResultContext, offset: usize, len: usize) {
            self.data.splice(offset..(offset + len), vec![0; len]);
        }
    }

    fn v<T>(size: usize, value: T) -> Vec<T>
    where
        T: Clone,
    {
        let mut v = Vec::with_capacity(size);
        v.resize(size, value);
        v
    }

    fn convert_events(allocator: &mut BatchBufferAllocator) -> Vec<BatchBufferAllocatorEvent> {
        let mut events = Vec::new();
        while let Some(event) = allocator.pop_event() {
            events.push(event);
        }
        events
    }

    #[test]
    fn test_basic_write() {
        let mut allocator = BatchBufferAllocator::default();
        let mut index_buffer = BatchBuffer::new(VecBuffer::new(10));
        let mut sprites = HashMap::new();
        let mut context = ResultContext;

        let p0 = allocator.allocate(5);
        sprites.insert(p0, 1);

        let allocator_len = allocator.len();
        while let Some(event) = allocator.pop_event() {
            match event {
                BatchBufferAllocatorEvent::Clear(pointer) => {
                    index_buffer.clear(&mut context, pointer);
                }
                BatchBufferAllocatorEvent::Write(pointer) => {
                    if index_buffer.len() < allocator_len {
                        index_buffer.resize(&mut context, allocator_len * 2);
                    }
                    let sprite = sprites.get(&pointer).unwrap();
                    index_buffer.write(&mut context, pointer, &v(pointer.len, *sprite));
                }
                BatchBufferAllocatorEvent::ReallocateAll { pairs } => {
                    for pair in pairs.iter() {
                        sprites.remove(&pair.from);
                    }
                    for (sprite, pointer) in pairs
                        .iter()
                        .filter_map(|pair| {
                            sprites.remove(&pair.from).map(|sprite| (sprite, pair.to))
                        })
                        .collect::<Vec<_>>()
                    {
                        sprites.insert(pointer, sprite);
                    }
                }
            }
        }

        assert_eq!(
            index_buffer.buffer.data[0..allocator.len()],
            [1, 1, 1, 1, 1]
        );

        let p1 = allocator.allocate(3);
        sprites.insert(p1, 2);

        let p2 = allocator.allocate(4);
        sprites.insert(p2, 3);

        let allocator_len = allocator.len();
        while let Some(event) = allocator.pop_event() {
            match event {
                BatchBufferAllocatorEvent::Clear(pointer) => {
                    index_buffer.clear(&mut context, pointer);
                }
                BatchBufferAllocatorEvent::Write(pointer) => {
                    if index_buffer.len() < allocator_len {
                        index_buffer.resize(&mut context, allocator_len * 2);
                    }
                    let sprite = sprites.get(&pointer).unwrap();
                    index_buffer.write(&mut context, pointer, &v(pointer.len, *sprite));
                }
                BatchBufferAllocatorEvent::ReallocateAll { pairs } => {
                    for pair in pairs.iter() {
                        sprites.remove(&pair.from);
                    }
                    for (sprite, pointer) in pairs
                        .iter()
                        .filter_map(|pair| {
                            sprites.remove(&pair.from).map(|sprite| (sprite, pair.to))
                        })
                        .collect::<Vec<_>>()
                    {
                        sprites.insert(pointer, sprite);
                    }
                }
            }
        }

        assert_eq!(
            index_buffer.buffer.data[0..allocator.len()],
            [1, 1, 1, 1, 1, 2, 2, 2, 3, 3, 3, 3]
        );
    }

    #[test]
    fn test_free() {
        let mut allocator = BatchBufferAllocator::default();

        let p0 = allocator.allocate(1); // 5
        let p1 = allocator.allocate(2); // 1
        let p2 = allocator.allocate(4); // 2
        let p3 = allocator.allocate(3); // 4
        let p4 = allocator.allocate(5); // 3

        assert_eq!(allocator.len(), 15);

        allocator.free(p1);

        assert_eq!(allocator.len(), 15);
        assert_eq!(
            allocator.pending_pointers.len(),
            1,
            "{:?}",
            allocator.pending_pointers
        );
        assert_eq!(
            allocator.pending_pointers_grouped_by_last.len(),
            1,
            "{:?}",
            allocator.pending_pointers_grouped_by_last
        );

        allocator.free(p2);

        assert_eq!(allocator.len(), 15);
        assert_eq!(
            allocator.pending_pointers.len(),
            1,
            "{:?}",
            allocator.pending_pointers
        );
        assert_eq!(
            allocator.pending_pointers_grouped_by_last.len(),
            1,
            "{:?}",
            allocator.pending_pointers_grouped_by_last
        );

        allocator.free(p4);

        assert_eq!(allocator.len(), 10);
        assert_eq!(
            allocator.pending_pointers.len(),
            1,
            "{:?}",
            allocator.pending_pointers
        );
        assert_eq!(
            allocator.pending_pointers_grouped_by_last.len(),
            1,
            "{:?}",
            allocator.pending_pointers_grouped_by_last
        );

        allocator.free(p3);

        assert_eq!(allocator.len(), 1);
        assert_eq!(
            allocator.pending_pointers.len(),
            0,
            "{:?}",
            allocator.pending_pointers
        );
        assert_eq!(
            allocator.pending_pointers_grouped_by_last.len(),
            0,
            "{:?}",
            allocator.pending_pointers_grouped_by_last
        );

        allocator.free(p0);

        assert_eq!(allocator.len(), 0);
        assert_eq!(
            allocator.pending_pointers.len(),
            0,
            "{:?}",
            allocator.pending_pointers
        );
        assert_eq!(
            allocator.pending_pointers_grouped_by_last.len(),
            0,
            "{:?}",
            allocator.pending_pointers_grouped_by_last
        );
    }

    #[test]
    fn test_allocate() {
        let mut allocator = BatchBufferAllocator::default();

        let p0 = allocator.allocate(3);
        assert_eq!(p0.first, 0);
        assert_eq!(p0.len, 3);

        let p1 = allocator.allocate(4);
        assert_eq!(p1.first, 3);
        assert_eq!(p1.len, 4);

        let p2 = allocator.allocate(5);
        assert_eq!(p2.first, 7);
        assert_eq!(p2.len, 5);

        assert_eq!(allocator.pointers.len(), 3);
        assert_eq!(allocator.len(), 12);

        insta::assert_debug_snapshot!(convert_events(&mut allocator));
    }

    #[test]
    fn test_reallocate() {
        let mut allocator = BatchBufferAllocator::default();
        let p0 = allocator.allocate(10);
        let p1 = allocator.allocate(5);
        let _p2 = allocator.allocate(15);
        let p1 = allocator.reallocate(p1, 10);
        assert_eq!(p1.first, 30);
        assert_eq!(p1.len, 10);

        let p0 = allocator.reallocate(p0, 5);
        assert_eq!(p0.first, 10);
        assert_eq!(p0.len, 5);
    }

    #[test]
    fn test_defragmentation() {
        let mut allocator = BatchBufferAllocator::default();
        let p0 = allocator.allocate(3);
        let p1 = allocator.allocate(3);
        let _p2 = allocator.allocate(3);
        let p3 = allocator.allocate(3);
        let _p4 = allocator.allocate(3);

        allocator.free(p0);
        allocator.free(p1);
        allocator.free(p3);

        convert_events(&mut allocator);

        allocator.defragmentation();
        assert_eq!(allocator.pointers.len(), 2);
        assert_eq!(allocator.pending_pointers.len(), 0);
        assert_eq!(allocator.pending_pointers_grouped_by_last.len(), 0);
        assert_eq!(allocator.len, 6);

        insta::assert_debug_snapshot!(convert_events(&mut allocator));

        assert_eq!(allocator.pointers.get(&0).unwrap().len, 3);
        assert_eq!(allocator.pointers.get(&3).unwrap().len, 3);
    }

    #[test]
    fn test_events() {
        let mut allocator = BatchBufferAllocator::default();

        let p0 = allocator.allocate(1); // 4
        let p1 = allocator.allocate(2); // 1
        let p2 = allocator.allocate(4); // 3
        let _p3 = allocator.allocate(3);
        let p4 = allocator.allocate(5); // 2

        insta::assert_debug_snapshot!(convert_events(&mut allocator));

        allocator.free(p1);

        insta::assert_debug_snapshot!(convert_events(&mut allocator));

        allocator.free(p4);

        insta::assert_debug_snapshot!(convert_events(&mut allocator));

        allocator.free(p2);

        insta::assert_debug_snapshot!(convert_events(&mut allocator));

        let _p0 = allocator.reallocate(p0, 5);

        insta::assert_debug_snapshot!(convert_events(&mut allocator));
    }

    #[test]
    fn test_sort() {
        let mut allocator = BatchBufferAllocator::default();

        // 5, 3, 4, 1
        let _p0 = allocator.allocate(5);
        let _p1 = allocator.allocate(3);
        let _p2 = allocator.allocate(4);
        let _p3 = allocator.allocate(1);

        allocator.sort_by(|p0, p1| p0.len.cmp(&p1.len)); // 1, 3, 4, 5

        insta::assert_debug_snapshot!(convert_events(&mut allocator));

        let p1 = BatchBufferPointer::new(1, 3);
        allocator.free(p1); // 1, _, 4, 5

        allocator.allocate(2); // 1, 2, 4, 5

        allocator.sort_by(|p0, p1| p0.len.cmp(&p1.len));

        insta::assert_debug_snapshot!(convert_events(&mut allocator));

        assert_eq!(allocator.len, 12);
    }
}
