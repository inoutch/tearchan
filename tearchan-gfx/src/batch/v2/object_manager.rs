use crate::batch::types::{BatchAttributeIndex, BatchTypeArray, BatchTypeTransform};
use crate::batch::v2::buffer::{
    BatchBufferAllocator, BatchBufferAllocatorEvent, BatchBufferPointer,
};
use crate::batch::v2::object::BatchObject;
use crate::batch::DEFAULT_ORDER;
use std::collections::{HashMap, HashSet, VecDeque};
use tearchan_util::id_manager::IdManager;

#[derive(Hash, Eq, PartialEq, Copy, Clone, Debug)]
pub struct BatchObjectId(u64);

impl BatchObjectId {
    fn next(&self) -> Self {
        BatchObjectId(self.0 + 1)
    }
}

#[derive(Debug)]
pub enum BatchObjectEvent {
    Add {
        id: BatchObjectId,
    },
    Remove {
        id: BatchObjectId,
    },
    Write {
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
    },
    Clear {
        pointer: BatchBufferPointer,
    },
    Resize {
        len: usize,
    },
}

pub struct BatchObjectManager {
    id_manager: IdManager<BatchObjectId>,
    objects: HashMap<BatchObjectId, BatchObject>,
    object_ids_grouped_by_pointer: HashMap<BatchBufferPointer, BatchObjectId>,
    objects_will_be_rewritten: HashSet<(BatchObjectId, BatchAttributeIndex)>,
    allocator: BatchBufferAllocator,
    events: VecDeque<BatchObjectEvent>,
    len: usize,
}

impl BatchObjectManager {
    pub fn new(len: usize) -> Self {
        BatchObjectManager {
            id_manager: IdManager::new(BatchObjectId(0), |id| id.next()),
            objects: HashMap::new(),
            objects_will_be_rewritten: HashSet::new(),
            object_ids_grouped_by_pointer: HashMap::new(),
            allocator: BatchBufferAllocator::default(),
            events: VecDeque::new(),
            len,
        }
    }

    pub fn pop_event(&mut self) -> Option<BatchObjectEvent> {
        while let Some(event) = self.allocator.pop_event() {
            match event {
                BatchBufferAllocatorEvent::Write(pointer) => {
                    let object_id = self.object_ids_grouped_by_pointer.get(&pointer).unwrap();
                    let object = self.objects.get(object_id).unwrap();
                    for (i, _) in object.data().iter().enumerate() {
                        self.events.push_back(BatchObjectEvent::Write {
                            id: *object_id,
                            attribute: i as BatchAttributeIndex,
                        });
                    }
                }
                BatchBufferAllocatorEvent::Clear(pointer) => {
                    self.events.push_back(BatchObjectEvent::Clear { pointer });
                }
                BatchBufferAllocatorEvent::Reallocate { from, to } => {
                    let object_id = *self.object_ids_grouped_by_pointer.get(&from).unwrap();
                    let object = self.objects.get_mut(&object_id).unwrap();
                    self.object_ids_grouped_by_pointer.remove(&object.pointer());

                    object.set_pointer(to);
                    self.object_ids_grouped_by_pointer.insert(to, object_id);

                    for i in 0u32..object.data().len() as u32 {
                        self.objects_will_be_rewritten.insert((object_id, i));
                    }
                }
            }
        }
        let event = self.events.pop_front()?;
        match &event {
            BatchObjectEvent::Add { .. } => {}
            BatchObjectEvent::Remove { .. } => {}
            BatchObjectEvent::Write { id, attribute } => {
                self.objects_will_be_rewritten.remove(&(*id, *attribute));
            }
            BatchObjectEvent::Clear { .. } => {}
            BatchObjectEvent::Resize { .. } => {}
        }
        Some(event)
    }

    pub fn clear_events(&mut self) {
        self.objects_will_be_rewritten.clear();
    }

    #[inline]
    pub fn get(&self, id: BatchObjectId) -> Option<&BatchObject> {
        self.objects.get(&id)
    }

    pub fn add(
        &mut self,
        data: Vec<BatchTypeArray>,
        len: usize,
        order: Option<i32>,
    ) -> BatchObjectId {
        for datum in data.iter() {
            assert_eq!(datum.len(), len);
        }

        let id = self.id_manager.gen();
        let order = order.unwrap_or(DEFAULT_ORDER);
        let transforms = vec![BatchTypeTransform::None; data.len()];
        let pointer = self.allocator.allocate(len);
        let object = BatchObject::new(pointer, data, transforms, order);

        if self.allocator.len() > self.len {
            self.len = self.allocator.len() * 2;
            self.events
                .push_back(BatchObjectEvent::Resize { len: self.len });
        }

        for i in 0u32..object.data().len() as u32 {
            self.objects_will_be_rewritten.insert((id, i));
        }
        self.events.push_back(BatchObjectEvent::Add { id });
        self.objects.insert(id, object);
        self.object_ids_grouped_by_pointer.insert(pointer, id);
        id
    }

    pub fn remove(&mut self, id: BatchObjectId) -> Option<BatchObject> {
        self.events.push_back(BatchObjectEvent::Remove { id });
        let object = self.objects.remove(&id)?;
        self.allocator.free(object.pointer());
        self.object_ids_grouped_by_pointer.remove(&object.pointer());
        for attribute in 0..object.data().len() {
            self.objects_will_be_rewritten
                .remove(&(id, attribute as u32));
        }
        Some(object)
    }

    pub fn transform(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        transform: BatchTypeTransform,
    ) {
        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        object.set_transform(attribute, transform);

        let key = (id, attribute);
        if self.objects_will_be_rewritten.get(&key).is_none() {
            self.events
                .push_back(BatchObjectEvent::Write { id, attribute });
            self.objects_will_be_rewritten.insert(key);
        }
    }

    pub fn rewrite(
        &mut self,
        id: BatchObjectId,
        attribute: BatchAttributeIndex,
        data: BatchTypeArray,
    ) {
        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        let prev_data = match object.data().get(attribute as usize) {
            Some(data) => data,
            None => return,
        };
        assert_eq!(data.len(), prev_data.len());

        object.set_data(attribute, data);
        let key = (id, attribute);
        if self.objects_will_be_rewritten.get(&key).is_none() {
            self.events
                .push_back(BatchObjectEvent::Write { id, attribute });
            self.objects_will_be_rewritten.insert(key);
        }
    }

    pub fn replace(&mut self, id: BatchObjectId, len: usize, data: Vec<BatchTypeArray>) {
        for datum in data.iter() {
            assert_eq!(datum.len(), len);
        }

        let object = match self.objects.get_mut(&id) {
            Some(object) => object,
            None => return,
        };
        if !object.data().is_empty() && object.data().get(0).unwrap().len() == len {
            for (i, data) in data.into_iter().enumerate() {
                self.rewrite(id, i as u32, data);
            }
            return;
        }

        self.object_ids_grouped_by_pointer.remove(&object.pointer());

        let new_pointer = self.allocator.reallocate(object.pointer(), len);
        object.set_pointer(new_pointer);
        self.object_ids_grouped_by_pointer.insert(new_pointer, id);

        for i in 0u32..object.data().len() as u32 {
            self.objects_will_be_rewritten.insert((id, i));
        }
    }

    pub fn allocator_len(&self) -> usize {
        self.allocator.len()
    }

    pub fn allocator_is_empty(&self) -> bool {
        self.allocator.is_empty()
    }
}

#[cfg(test)]
mod test {
    use crate::batch::types::{BatchTypeArray, BatchTypeTransform};
    use crate::batch::v2::object_manager::{BatchObjectEvent, BatchObjectManager};
    use nalgebra_glm::{vec2, vec3, Mat2, Mat3};

    fn convert_events(manager: &mut BatchObjectManager) -> Vec<BatchObjectEvent> {
        let mut events = Vec::new();
        while let Some(event) = manager.pop_event() {
            events.push(event);
        }
        events
    }

    #[test]
    fn test_valid_len() {
        let mut manager = BatchObjectManager::new(100);
        manager.add(
            vec![
                BatchTypeArray::V1F32 { data: vec![0.0f32] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(0.0f32, 0.0f32)],
                },
            ],
            1,
            None,
        );
    }

    #[test]
    #[should_panic]
    fn test_invalid_len() {
        let mut manager = BatchObjectManager::new(100);
        manager.add(
            vec![
                BatchTypeArray::V1F32 {
                    data: vec![0.0f32, 0.0f32],
                },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(0.0f32, 0.0f32)],
                },
            ],
            2,
            None,
        );
    }

    #[test]
    fn test_cleanup() {
        let mut manager = BatchObjectManager::new(100);
        let data = vec![
            BatchTypeArray::V1F32 { data: vec![0.0f32] },
            BatchTypeArray::V2F32 {
                data: vec![vec2(0.0f32, 0.0f32)],
            },
        ];
        let id0 = manager.add(data, 1, None);

        let data = vec![
            BatchTypeArray::V1F32 { data: vec![0.0f32] },
            BatchTypeArray::V2F32 {
                data: vec![vec2(0.0f32, 0.0f32)],
            },
        ];
        let id1 = manager.add(data, 1, None);
        manager.remove(id0);
        manager.remove(id1);

        assert_eq!(manager.objects.len(), 0);
        assert_eq!(manager.object_ids_grouped_by_pointer.len(), 0);
        assert_eq!(manager.objects_will_be_rewritten.len(), 0);
        assert_eq!(manager.allocator.len(), 0);
    }

    #[test]
    fn test_events() {
        let mut manager = BatchObjectManager::new(100);
        let id0 = manager.add(
            vec![
                BatchTypeArray::V1F32 { data: vec![1.0f32] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(1.0f32, 1.0f32)],
                },
            ],
            1,
            None,
        );
        let id1 = manager.add(
            vec![
                BatchTypeArray::V1F32 { data: vec![2.0f32] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(2.0f32, 2.0f32)],
                },
            ],
            1,
            None,
        );

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        manager.transform(
            id0,
            1,
            BatchTypeTransform::Mat3F32 {
                m: Mat3::new(
                    1.0f32, 0.0f32, 0.0f32, 0.0f32, 1.0f32, 0.0f32, 0.0f32, 0.0f32, 1.0f32,
                ),
            },
        );
        manager.transform(
            id1,
            0,
            BatchTypeTransform::Mat2F32 {
                m: Mat2::new(1.0f32, 0.0f32, 0.0f32, 1.0f32),
            },
        );

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        manager.remove(id0);

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        manager.remove(id1);

        insta::assert_debug_snapshot!(convert_events(&mut manager));
    }

    #[test]
    fn test_resize_event() {
        let mut manager = BatchObjectManager::new(2);
        let _id0 = manager.add(
            vec![
                BatchTypeArray::V1U32 { data: vec![1, 1] },
                BatchTypeArray::V2F32 {
                    data: vec![vec2(1.0f32, 1.0f32), vec2(1.0f32, 1.0f32)],
                },
                BatchTypeArray::V3F32 {
                    data: vec![vec3(1.0f32, 1.0f32, 1.0f32), vec3(1.0f32, 1.0f32, 1.0f32)],
                },
            ],
            2,
            None,
        );

        insta::assert_debug_snapshot!(convert_events(&mut manager));

        let _id1 = manager.add(
            vec![
                BatchTypeArray::V1U32 {
                    data: vec![2, 2, 2],
                },
                BatchTypeArray::V2F32 {
                    data: vec![
                        vec2(2.0f32, 2.0f32),
                        vec2(2.0f32, 2.0f32),
                        vec2(2.0f32, 2.0f32),
                    ],
                },
                BatchTypeArray::V3F32 {
                    data: vec![
                        vec3(2.0f32, 2.0f32, 2.0f32),
                        vec3(2.0f32, 2.0f32, 2.0f32),
                        vec3(2.0f32, 2.0f32, 2.0f32),
                    ],
                },
            ],
            3,
            None,
        );
        insta::assert_debug_snapshot!(convert_events(&mut manager));
    }
}
