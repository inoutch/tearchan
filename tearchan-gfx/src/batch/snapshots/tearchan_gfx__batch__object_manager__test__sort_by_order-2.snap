---
source: tearchan-gfx/src/batch/v2/object_manager.rs
expression: convert_events(&mut manager)

---
[
    Add {
        id: BatchObjectId(
            3,
        ),
    },
    ResizeIndexBuffer {
        len: 10,
    },
    ResizeVertexBuffer {
        len: 20,
    },
    Add {
        id: BatchObjectId(
            4,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            1,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            4,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            2,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            3,
        ),
    },
    WriteToIndexBuffer {
        id: BatchObjectId(
            0,
        ),
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            3,
        ),
        attribute: 0,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            3,
        ),
        attribute: 1,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            3,
        ),
        attribute: 2,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            4,
        ),
        attribute: 0,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            4,
        ),
        attribute: 1,
    },
    WriteToVertexBuffer {
        id: BatchObjectId(
            4,
        ),
        attribute: 2,
    },
]
